### PR TITLE
port masked_select from TH to ATen and optimize perf on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
This PR ports `masked_select` from TH to ATen and optimize the performance on CPU with TensorIterator.

https://github.com/pytorch/pytorch/issues/33053

1. single socket run: up to **5.4x** speedup;
2. single core run: up to **1.16x** speedup.

### CODE REVIEW

**Comment by mingfeima** [comment]
In order to parallel `masked_select` with TensorIterator, extra meta data which records the `offset` of output tensor is needed, `masked_prefix_sum`. It is calculated with **std::partial_sum** at the moment since PyTorch supports C++14. Once PyTorch upgrades to C++17, we can use **std::exclusive_scan** which is from parallel STL, thus have better performance on CPU. I checked CUDA implementation has used **thrust::exclusive_scan** but CPU counterpart has dependency on C++17.

Below shows the performance compare between **original** and **this pr**, benchmark code available at this [op_bench-py](https://github.com/mingfeima/op_bench-py), to reproduce, `./run.sh masked_select.py`. Test machine is Xeon Skylake 6148 with 2*20 cores @2.40GHz.

The unit is `ms per iteration`, the lower the better.
The benchmark takes `mask = input.ge(0.0)`, so roughly **50%** of the input elements are selected.

#### single socket run

input size | original | this pr | speedup
-- | -- | -- | --
[128 1000] | 0.817 | 0.295 | 2.77
[256 1000] | 1.899 | 0.413 | 4.60
[512 1000] | 3.747 | 0.700 | 5.35
[1024 1000] | 6.696 | 1.238 | 5.41

#### single core run

input size | original | this pr | speedup
-- | -- | -- | --
[128 1000] | 0.812 | 0.703 | 1.16
[256 1000] | 1.620 | 1.421 | 1.14
[512 1000] | 3.246 | 2.823 | 1.15
[1024 1000] | 6.488 | 5.675 | 1.14

File: aten/src/ATen/native/cpu/IndexKernel.cpp
Comment (mingfeima) [inline aten/src/ATen/native/cpu/IndexKernel.cpp]: `std::partial_sum` functionally equals to `std::inclusive_scan` so we have to make adjustment here.
Once we are able to use `std::exclusive_scan`, this line needs to be:
`int64_t offset_bytes = offset * sizeof(scalar_t);`

**Comment by mingfeima** [comment]
rebased

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (mingfeima) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Makes sense! By doing this, i also need to modify `test_masked_select` in `test_torch.py` a little bit. See [L10835](https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L10835) it assumes that `masked_select` and `masked_select_out` generates only 1 warning message.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (mingfeima) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: `cumsum` now runs in sequential so performance not very good. I have another colleague working on `cumsum` (see #33280 ), it's almost finished but even with that `cumsum` is still sequential on a flat tensor with `dim=0`.

`std::partial_sum` from STL has slightly better perf on CPU than `cumsum`. Anyway, the diff is not very big, if you feel like a unified code path for both CUDA and CPU, i assume use `cumsum` is also acceptable. But I was wondering that [thrust::exclusive_scan](https://docs.nvidia.com/cuda/thrust/index.html#prefix-sums) should promise a good performance on CUDA, is there a problem with that?

```bash
#use std::partial_sum
input size: [128 1000]; output size: [63941]: time = 0.295 ms
input size: [256 1000]; output size: [127483]: time = 0.413 ms
input size: [512 1000]; output size: [255929]: time = 0.700 ms
input size: [1024 1000]; output size: [511179]: time = 1.238 ms

# use cumsum
input size: [128 1000]; output size: [63941]: time = 0.330 ms
input size: [256 1000]; output size: [127483]: time = 0.482 ms
input size: [512 1000]; output size: [255929]: time = 0.729 ms
input size: [1024 1000]; output size: [511179]: time = 1.283 ms
```

Comment (kurtamohler) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: If the extra bit of performance that `std::partial_sum` gives is important to you, then it's alright with me. Maybe someone from Facebook should weigh in though.

However, it seems like it would be ideal if we could make cumsum use `std::partial_sum` underneath so that it gives the similar performance to what you have. Also, do you happen to know what the performance would be if you use @XiaobingSuper 's updated `cumsum`?

**Comment by VitalyFedyunin** [comment]
```
Feb 24 08:29:38 ======================================================================
Feb 24 08:29:38 ERROR [0.008s]: test_fmod_cpu_float32 (__main__.TestTorchDeviceTypeCPU)
Feb 24 08:29:38 ----------------------------------------------------------------------
Feb 24 08:29:38 Traceback (most recent call last):
Feb 24 08:29:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 197, in instantiated_test
Feb 24 08:29:38     result = test(self, device_arg, dtype)
Feb 24 08:29:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 398, in only_fn
Feb 24 08:29:38     return fn(slf, device, *args, **kwargs)
Feb 24 08:29:38   File "test_torch.py", line 13577, in test_fmod
Feb 24 08:29:38     _test_fmod_with_size_tensor(size, device, dtype)
Feb 24 08:29:38   File "test_torch.py", line 13556, in _test_fmod_with_size_tensor
Feb 24 08:29:38     a = torch.rand(size=size, device=device, dtype=dtype)
Feb 24 08:29:38 RuntimeError: _th_uniform_ not supported on CPUType for Byte
```

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (mingfeima) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: That makes sense. I think you can submit a new PR with an unified CPU and CUDA implementation to replace this one :)

**Comment by mingfeima** [comment]
> test_fmod_cpu_float32

Sorry, it is this related to this one?

**Comment by VitalyFedyunin** [comment]
> > test_fmod_cpu_float32
> 
> Sorry, it is this related to this one?

Oops, wrong browser tab, sorry

**Comment by kurtamohler** [comment]
Hey @mingfeima , are you sure your performance comparison between your implementation and the old TH implementation is correct? I was running some measurements on my cuda implementation, and I decided to collect the cpu performance data as well to make sure my changes won't negatively affect yours. Here's the performance I see with a single CPU core:

```
dtype | tensor_size | mask_true_ratio | TH CPU time | ATen cpu (mingfeima) time | speedup mingfeima ATen over TH
-- | -- | -- | -- | -- | --
torch.float64 | 1000 | 0 | 0.0000146 | 0.0001218 | 0.1198686371
torch.float64 | 1000 | 0.25 | 0.0000182 | 0.0001524 | 0.1194225722
torch.float64 | 1000 | 0.5 | 0.0000176 | 0.0001544 | 0.1139896373
torch.float64 | 1000 | 0.75 | 0.0000174 | 0.0001538 | 0.1131339402
torch.float64 | 1000 | 1 | 0.0000172 | 0.0001504 | 0.1143617021
torch.float64 | 10000 | 0 | 0.0000524 | 0.0006164 | 0.08500973394
torch.float64 | 10000 | 0.25 | 0.0000722 | 0.0007122 | 0.101376018
torch.float64 | 10000 | 0.5 | 0.000076 | 0.0007412 | 0.1025364274
torch.float64 | 10000 | 0.75 | 0.000072 | 0.000738 | 0.09756097561
torch.float64 | 10000 | 1 | 0.0000586 | 0.0007264 | 0.08067180617
torch.float64 | 100000 | 0 | 0.000439 | 0.0055718 | 0.07878961915
torch.float64 | 100000 | 0.25 | 0.0006324 | 0.006367 | 0.09932464269
torch.float64 | 100000 | 0.5 | 0.0007926 | 0.0067212 | 0.1179253705
torch.float64 | 100000 | 0.75 | 0.0006554 | 0.0066458 | 0.09861867646
torch.float64 | 100000 | 1 | 0.0004654 | 0.0065112 | 0.07147683991
torch.float64 | 1000000 | 0 | 0.0042768 | 0.0559348 | 0.07646045038
torch.float64 | 1000000 | 0.25 | 0.0063654 | 0.064207 | 0.09913872319
torch.float64 | 1000000 | 0.5 | 0.0083026 | 0.0674632 | 0.1230685766
torch.float64 | 1000000 | 0.75 | 0.0067032 | 0.0665706 | 0.1006930988
torch.float64 | 1000000 | 1 | 0.0050948 | 0.0653446 | 0.07796818712
torch.float64 | 10000000 | 0 | 0.0424076 | 0.577658 | 0.07341298831
torch.float64 | 10000000 | 0.25 | 0.0647738 | 0.6789826 | 0.09539832096
torch.float64 | 10000000 | 0.5 | 0.0916282 | 0.7228486 | 0.1267598775
torch.float64 | 10000000 | 0.75 | 0.0796012 | 0.7180316 | 0.110860302
torch.float64 | 10000000 | 1 | 0.0646096 | 0.7097978 | 0.09102535962
```

My performance measurement script is here: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/masked_select/masked_select-perf.py

**Comment by kurtamohler** [comment]
Actually, I think your performance measurement script is sufficient. When I run it, I get something that seems to be similar to my measurements, showing a performance decrease.

TH CPU masked_select:
```
input size: [128 1000]; output size: [63941]: time = 1.027 ms
input size: [256 1000]; output size: [127483]: time = 2.225 ms
input size: [512 1000]; output size: [255929]: time = 4.103 ms
input size: [1024 1000]; output size: [511179]: time = 7.888 ms
```

ATen CPU masked_select:
```
input size: [128 1000]; output size: [63941]: time = 5.091 ms
input size: [256 1000]; output size: [127483]: time = 6.737 ms
input size: [512 1000]; output size: [255929]: time = 10.314 ms
input size: [1024 1000]; output size: [511179]: time = 17.129 ms
```



TH CPU masked_select (modified to use only a single core):
```
input size: [128 1000]; output size: [63941]: time = 1.030 ms
input size: [256 1000]; output size: [127483]: time = 2.098 ms
input size: [512 1000]; output size: [255929]: time = 4.241 ms
input size: [1024 1000]; output size: [511179]: time = 8.834 ms
```

ATen CPU masked_select (modified to use only a single core):
```
input size: [128 1000]; output size: [63941]: time = 8.616 ms
input size: [256 1000]; output size: [127483]: time = 17.228 ms
input size: [512 1000]; output size: [255929]: time = 34.527 ms
input size: [1024 1000]; output size: [511179]: time = 69.079 ms
```

**Comment by mingfeima** [comment]
@kurtamohler What type of CPU are you using? And how about the environment setting?

**Comment by mingfeima** [comment]
Post my machine configuration:
```bash
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                80
On-line CPU(s) list:   0-79
Thread(s) per core:    2
Core(s) per socket:    20
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz
Stepping:              4
CPU MHz:               1000.125
CPU max MHz:           3700.0000
CPU min MHz:           1000.0000
BogoMIPS:              4800.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              28160K
NUMA node0 CPU(s):     0-19,40-59
NUMA node1 CPU(s):     20-39,60-79
```

* original log
```bash
### this refers to ATen implementation
(pytorch-mingfei) [mingfeim@mlt-skx090 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1
### using numactl --physcpubind=0-19 --membind=0


### using OMP_NUM_THREADS=20
input size: [128 1000]; output size: [63941]: time = 0.295 ms
input size: [256 1000]; output size: [127483]: time = 0.413 ms
input size: [512 1000]; output size: [255929]: time = 0.700 ms
input size: [1024 1000]; output size: [511179]: time = 1.238 ms

### using OMP_NUM_THREADS=1
input size: [128 1000]; output size: [63941]: time = 0.703 ms
input size: [256 1000]; output size: [127483]: time = 1.421 ms
input size: [512 1000]; output size: [255929]: time = 2.823 ms
input size: [1024 1000]; output size: [511179]: time = 5.675 ms

### this refers to original TH implementation
(pytorch-cuda) [mingfeim@mlt-skx090 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1
### using numactl --physcpubind=0-19 --membind=0


### using OMP_NUM_THREADS=20
input size: [128 1000]; output size: [63941]: time = 0.817 ms
input size: [256 1000]; output size: [127483]: time = 1.899 ms
input size: [512 1000]; output size: [255929]: time = 3.747 ms
input size: [1024 1000]; output size: [511179]: time = 6.696 ms

### using OMP_NUM_THREADS=1
input size: [128 1000]; output size: [63941]: time = 0.812 ms
input size: [256 1000]; output size: [127483]: time = 1.620 ms
input size: [512 1000]; output size: [255929]: time = 3.246 ms
input size: [1024 1000]; output size: [511179]: time = 6.488 ms
```

* this is log when update benchmark script with only two `time.time()` calls, to verify the overhead of `time.time()`. Usually the minimal time segment should ms level for benchmark purpose. The variance here is more related to python rather than time().
```bash
### this refers to ATen implementation
(pytorch-mingfei) [mingfeim@mlt-skx089 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1
### using numactl --physcpubind=0-19 --membind=0


### using OMP_NUM_THREADS=20
input size: [128 1000]; output size: [63941]: time = 0.320 ms
input size: [256 1000]; output size: [127483]: time = 0.447 ms
input size: [512 1000]; output size: [255929]: time = 0.706 ms
input size: [1024 1000]; output size: [511179]: time = 1.255 ms

### using OMP_NUM_THREADS=1
input size: [128 1000]; output size: [63941]: time = 0.698 ms
input size: [256 1000]; output size: [127483]: time = 1.416 ms
input size: [512 1000]; output size: [255929]: time = 2.814 ms
input size: [1024 1000]; output size: [511179]: time = 5.651 ms


### this refers to original TH implementation
(pytorch-cuda) [mingfeim@mlt-skx089 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1
### using numactl --physcpubind=0-19 --membind=0


### using OMP_NUM_THREADS=20
input size: [128 1000]; output size: [63941]: time = 0.874 ms
input size: [256 1000]; output size: [127483]: time = 1.900 ms
input size: [512 1000]; output size: [255929]: time = 3.711 ms
input size: [1024 1000]; output size: [511179]: time = 6.727 ms

### using OMP_NUM_THREADS=1
input size: [128 1000]; output size: [63941]: time = 0.860 ms
input size: [256 1000]; output size: [127483]: time = 1.612 ms
input size: [512 1000]; output size: [255929]: time = 3.228 ms
input size: [1024 1000]; output size: [511179]: time = 6.450 ms
```

From the algorithm level, it is possible that on single core with larger tensor, this ATen implementation might introduce a downgrade. Since `prefix-sum` requires additional memory footprint thus more demand on memory bandwidth, in case L1/L2/LLC fails to hold it, you will see a drop.

More efficient way here is to write two sets of kernels for `sequential` and `parallel`, the `parallel` one goes like this ( I actually copied the TH cuda implementation). And `sequential` one could use `nonzero` to mark the offset of `input` tensor and feed TensorIterator with a re-strided input (instead of re-strided output), therefore numel of TensorIteraor will be only the TRUEs in the mask.
However this is not very perf critical op and also to avoid complexity in testing, i did not do it (so as the cuda code).

**Comment by kurtamohler** [comment]
My cpu is an AMD Ryzen Threadripper 2970WX, which has 32 KB L1d and 512 KB L2 per core, and 8 MB per 3 cores. But I don't think the difference between our cache sizes could be the explanation, because the measurements I posted show this slowdown even in relatively small tensor sizes, down to 1000, which should fit within L1 for both of us.

I'm not really sure what jemalloc is. I guess it's just one of the implementations of malloc available as a kernel module, right? I'm guessing my malloc is not jemalloc, which might explain our machines' performance difference?

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (kurtamohler) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp]: While working on my cuda version of masked_select, I realized that we can reuse the final element from the cumulative sum result to get `numel`, rather than running a whole other sum calculation.

**Comment by kurtamohler** [comment]
@VitalyFedyunin , when you get a chance, could you review this PR? I've ported the CUDA version, but it depends on some of mingfeima's changes here, so I think I have to wait until this PR is merged before creating my PR.

**Comment by anjali411** [comment]
@mingfeima thanks for the PR! I think you need to rebase to resolve the merge conflicts and then should be good to go. 
@VitalyFedyunin can you do a final review for this PR?

**Review by anjali411 — APPROVED** [review]

**Review by anjali411 — CHANGES_REQUESTED** [review]
@mingfeima do you have any updates on the CPU performance issue?
@kurtamohler would you like to work on this after you finish the TH to ATen port for cuda?

**Comment by kurtamohler** [comment]
@anjali411 , you mean work on the performance issue specifically? Sure, I can look into what's causing it.

**Comment by mingfeima** [comment]
Updates on my side with latest code base.

I switched to high-end Xeon Skylake CPU 8180 (2 sockets, 56 cores, 2.5GHz), the result is similar to Xeon Skylake 6148 (2 sockets, 40 cores, 2.4GHz)

```bash
### this is official pytorch 1.4 release
(pytorch-cuda) [mingfeim@mlt-skx052 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1


### using OMP_NUM_THREADS=28
### using numactl --physcpubind=0-27 --membind=0

input size: [128 1000]; output size: [63941]: time = 0.820 ms
input size: [256 1000]; output size: [127483]: time = 1.841 ms
input size: [512 1000]; output size: [255929]: time = 3.624 ms
input size: [1024 1000]; output size: [511179]: time = 6.668 ms

### using OMP_NUM_THREADS=1
### using numactl --physcpubind=0 --membind=0

input size: [128 1000]; output size: [63941]: time = 0.955 ms
input size: [256 1000]; output size: [127483]: time = 1.891 ms
input size: [512 1000]; output size: [255929]: time = 3.746 ms
input size: [1024 1000]; output size: [511179]: time = 7.508 ms


### this is the latest code base in this PR
(pytorch-mingfei) [mingfeim@mlt-skx052 op_bench-py]$ ./run.sh masked_select.py

### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1


### using OMP_NUM_THREADS=28
### using numactl --physcpubind=0-27 --membind=0

input size: [128 1000]; output size: [63941]: time = 0.326 ms
input size: [256 1000]; output size: [127483]: time = 0.451 ms
input size: [512 1000]; output size: [255929]: time = 0.699 ms
input size: [1024 1000]; output size: [511179]: time = 1.220 ms

### using OMP_NUM_THREADS=1
### using numactl --physcpubind=0 --membind=0

input size: [128 1000]; output size: [63941]: time = 1.285 ms
input size: [256 1000]; output size: [127483]: time = 2.363 ms
input size: [512 1000]; output size: [255929]: time = 4.045 ms
input size: [1024 1000]; output size: [511179]: time = 7.782 ms
```

1. We don't have AMD threaddripper series on site so i can't reproduce @kurtamohler 's result. But we do have AMD server CPUs (EPYC 7742 rome), will this one do the job?
2. The current implementation is a trade off, in order to parallel, it computes some meta data (due to the semantics) so it is possible that it gets slower on some particular configuration. @anjali411 do you have any specific configurations (input tensor size, mask tensor density, thread number, etc.) for masked_select? We don't have any key models constrained by `masked_select` so i just randomly pick up some configs in the benchmark.

**Comment by kurtamohler** [comment]
It's interesting that your PR is getting a little bit of a single-core performance slowdown when you use your new Xeon 8180, whereas your previous Xeon 6148 measurements showed a single-core speedup. But this slowdown is pretty small compared to what I was getting with the Threadripper. I have no idea how comparable the EPYC 7742 rome is, but it would be worth finding out what kind of performance it gives if you have some time to measure it.

**Comment by anjali411** [comment]
> Updates on my side with latest code base.
> 
> I switched to high-end Xeon Skylake CPU 8180 (2 sockets, 56 cores, 2.5GHz), the result is similar to Xeon Skylake 6148 (2 sockets, 40 cores, 2.4GHz)
> 
> ```shell
> ### this is official pytorch 1.4 release
> (pytorch-cuda) [mingfeim@mlt-skx052 op_bench-py]$ ./run.sh masked_select.py
> 
> ### using KMP_AFFINITY=granularity=fine,compact,1,0
> ### using KMP_BLOCKTIME=1
> 
> 
> ### using OMP_NUM_THREADS=28
> ### using numactl --physcpubind=0-27 --membind=0
> 
> input size: [128 1000]; output size: [63941]: time = 0.820 ms
> input size: [256 1000]; output size: [127483]: time = 1.841 ms
> input size: [512 1000]; output size: [255929]: time = 3.624 ms
> input size: [1024 1000]; output size: [511179]: time = 6.668 ms
> 
> ### using OMP_NUM_THREADS=1
> ### using numactl --physcpubind=0 --membind=0
> 
> input size: [128 1000]; output size: [63941]: time = 0.955 ms
> input size: [256 1000]; output size: [127483]: time = 1.891 ms
> input size: [512 1000]; output size: [255929]: time = 3.746 ms
> input size: [1024 1000]; output size: [511179]: time = 7.508 ms
> 
> 
> ### this is the latest code base in this PR
> (pytorch-mingfei) [mingfeim@mlt-skx052 op_bench-py]$ ./run.sh masked_select.py
> 
> ### using KMP_AFFINITY=granularity=fine,compact,1,0
> ### using KMP_BLOCKTIME=1
> 
> 
> ### using OMP_NUM_THREADS=28
> ### using numactl --physcpubind=0-27 --membind=0
> 
> input size: [128 1000]; output size: [63941]: time = 0.326 ms
> input size: [256 1000]; output size: [127483]: time = 0.451 ms
> input size: [512 1000]; output size: [255929]: time = 0.699 ms
> input size: [1024 1000]; output size: [511179]: time = 1.220 ms
> 
> ### using OMP_NUM_THREADS=1
> ### using numactl --physcpubind=0 --membind=0
> 
> input size: [128 1000]; output size: [63941]: time = 1.285 ms
> input size: [256 1000]; output size: [127483]: time = 2.363 ms
> input size: [512 1000]; output size: [255929]: time = 4.045 ms
> input size: [1024 1000]; output size: [511179]: time = 7.782 ms
> ```
> 
> 1. We don't have AMD threaddripper series on site so i can't reproduce @kurtamohler 's result. But we do have AMD server CPUs (EPYC 7742 rome), will this one do the job?
> 2. The current implementation is a trade off, in order to parallel, it computes some meta data (due to the semantics) so it is possible that it gets slower on some particular configuration. @anjali411 do you have any specific configurations (input tensor size, mask tensor density, thread number, etc.) for masked_select? We don't have any key models constrained by `masked_select` so i just randomly pick up some configs in the benchmark.

@mingfeima thanks for your reply. I am not entirely sure.

@ngimel do we have any specific config for which we should test the masked select performance? are there any models that will be affected by this?

**Comment by ngimel** [comment]
We've synced internally, and given unresolved AMD perf concerns and single-threaded perf regressions, we'd like to either
a) address those
b) or if this cannot be done easily, do a dumb port of current implementation to ATen, preferably using TensorIterator, to unblock complex type work that depends on masked_select, and do perf improvements later. 
@mingfeima let us know if this is feasible, or if you prefer option b) to be done by someone else.

**Comment by mingfeima** [comment]
@ngimel sure, working on both issues.

**Comment by mingfeima** [comment]
@kurtamohler hi, I have add a serial kernel which has pretty much the same implementation as old TH code. Could you please give it a try? You could change this line [here](https://github.com/pytorch/pytorch/pull/33269/files#diff-5312d553f68ae8e12304275aab5ac43cR659) to let operator always call the serial kernel, thus we will have a glue to whether the serial kernel has similar performance with old TH code on your machine. Thanks!

**Comment by kurtamohler** [comment]
@mingfeima , I tried setting `use_serial_kernel = true`, but that didn't change the performance very much.

I decided to do a couple py-spy runs to profile the original version of masked_select and your version. Turns out that about 85% of the runtime within your version of masked_select is spent in the `at::sum()` call here: https://github.com/pytorch/pytorch/pull/33269/files#diff-5312d553f68ae8e12304275aab5ac43cR647

I put the py-spy flamegraphs in this zip file if you want to check them out. You can open the svg files in a web browser: 
[masked_select_cpu_py-spy_flamegraphs.zip](https://github.com/pytorch/pytorch/files/4512676/masked_select_cpu_py-spy_flamegraphs.zip)

**Comment by ngimel** [comment]
Thanks for investigating, @kurtamohler! Can you please quickly check the performance of .sum() for bool, byte and say float tensors? The `mask` is currently bool, right? If for some reason `.sum()` on bool is currently very slow on AMD processors that would be interesting to know.

**Comment by kurtamohler** [comment]
Here's what I get with this performance measurement script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/sum/sum_perf.py

device | data type | tensor size | torch.sum() time, seconds
-- | -- | -- | --
cpu | torch.float32 | 1000 | 0.000048
cpu | torch.float32 | 10000 | 0.000079
cpu | torch.float32 | 100000 | 0.000383
cpu | torch.float32 | 1000000 | 0.003521
cpu | torch.float32 | 10000000 | 0.037511
cpu | torch.int32 | 1000 | 0.000129
cpu | torch.int32 | 10000 | 0.000655
cpu | torch.int32 | 100000 | 0.005942
cpu | torch.int32 | 1000000 | 0.059008
cpu | torch.int32 | 10000000 | 0.610154
cpu | torch.bool | 1000 | 0.000129
cpu | torch.bool | 10000 | 0.000655
cpu | torch.bool | 100000 | 0.005943
cpu | torch.bool | 1000000 | 0.059016
cpu | torch.bool | 10000000 | 0.610713
cpu | torch.uint8 | 1000 | 0.000127
cpu | torch.uint8 | 10000 | 0.000642
cpu | torch.uint8 | 100000 | 0.005802
cpu | torch.uint8 | 1000000 | 0.05755
cpu | torch.uint8 | 10000000 | 0.596991

Interestingly, summing a tensor of floats is up to 16 times faster than summing bools, ints, or uint8's on AMD Threadripper. That's definitely a surprise to me. I think this means that there's nothing wrong with @mingfeima 's masked_select implementation, and we should look into what's wrong with torch.sum(). I'd be happy to investigate further.

**Comment by mingfeima** [comment]
@ngimel could you please review this PR?

**Comment by ngimel** [comment]
as far as I understand, the problem of `_mask.sum()` performance and thus overall slowdown is still not resolved?

**Comment by mingfeima** [comment]
@ngimel `_mask.sum()` performance issue is reported on AMD threadripper platform.
I tried to double check this issue but we only have AMD server CPUs on which the performance data is just as normal as Intel Xeon.

Below shows the performance compare between this pr and official pytorch 1.4 on AMD EPYC 7742, 2 socket, 64 core per socket, 8 numa nodes (16 core per numa node), 3.4 GHz.

thispr |   |   | 1.4 |   |  
-- | -- | -- | -- | -- | --
omp | input | time/ms | omp | input | time/ms
16 | 128 | 0.62 | 16 | 128 | 0.725
16 | 256 | 0.743 | 16 | 256 | 1.431
16 | 512 | 1.386 | 16 | 512 | 2.813
16 | 1024 | 2.687 | 16 | 1024 | 5.529
1 | 128 | 0.77 | 1 | 128 | 0.699
1 | 256 | 1.525 | 1 | 256 | 1.382
1 | 512 | 3.043 | 1 | 512 | 2.749
1 | 1024 | 6.337 | 1 | 1024 | 5.479


@kurtamohler do you have any update on the `sum()` performance issue?

**Comment by kurtamohler** [comment]
@mingfeima , I haven't had time to work on this for the past week or two, but the last thing I found is that the `at::to()` operation is the bottleneck. I made a benchmark script that shows the Threadripper gets about 100 times slower performance for `at::to()` than my Intel laptop. https://github.com/pytorch/pytorch/issues/37053#issuecomment-622147971

**Comment by kurtamohler** [comment]
It turns out that the performance decrease was my fault because I was building pytorch with `DEBUG=1`. Without that flag, I get good performance.

I had assumed that comparing the `DEBUG=1` builds of both the TH masked_select and your ATen port would be acceptable, but it turns out that the kernel for the `at::to()` call inside of `at::sum()` has quite a few layers of abstraction that don't get optimized away when `DEBUG=1`. But without that flag, optimization speeds up `at::sum()` to be comparable to the TH summation.

**Comment by ngimel** [comment]
Ok, cool, so AMD perf issues are resolved.

**Review by ngimel — APPROVED** [review]
Generally looks good, just some small comments.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (mingfeima) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: solved!

File: aten/src/ATen/native/cpu/IndexKernel.cpp
Comment (mingfeima) [reply aten/src/ATen/native/cpu/IndexKernel.cpp]: yes, strides[0] is always 0, fixed.

File: aten/src/ATen/native/cpu/IndexKernel.cpp
Comment (mingfeima) [reply aten/src/ATen/native/cpu/IndexKernel.cpp]: fixed!

**Comment by mingfeima** [comment]
@ngimel thanks for the comments, all fixed!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**
